### PR TITLE
Migrate Travis to container-based infrastructure

### PR DESCRIPTION
**Pull request by ximinez**
* http://docs.travis-ci.com/user/migrating-from-legacy/?utm_source=legacy-notice&utm_medium=banner&utm_campaign=legacy-upgrade

Be sure to check the Travis CI results at the bottom of the PR. That's what shows that it really works.

Reviewers: @vinniefalco @rec

### CODE REVIEW

**Comment by ximinez** [comment]
I'll note that the Travis unity builds have taken over 30 minutes, and the nonunity builds are still running. That's because they're building Boost (independently, unfortunately). However, the results of that build are cached, so subsequent builds will be faster. *I don't know if the caches persist across PRs and branches*, so that is a potential risk of this PR.

**Comment by ximinez** [comment]
Caching behavior should improve once this is merged or released. http://docs.travis-ci.com/user/caching/#Fetching-and-storing-caches

File: .travis.yml [hunk 1]
```diff
@@ -1,34 +1,45 @@
...
   - clang
   - gcc
 env:
-  - TARGET=debug
-  - TARGET=debug.nounity
-  # We can specify any combination of builds here, for example, to
-  # include release builds, too, uncomment the following lines.
-  #- TARGET=release
-  #- TARGET=release.nounity
+  global:
```
Comment (rec) [inline .travis.yml]: I love this!

This is getting a little long, perhaps add whitespace between sections like this example:  https://github.com/rec/echomesh/blob/master/code/python/echomesh/settings/settings.yml

File: .travis.yml [hunk 2]
```diff
@@ -1,34 +1,45 @@
...
+    #- TARGET=release
+    #- TARGET=release.nounity
+addons:
+  apt:
+    sources:
+    - ubuntu-toolchain-r-test
+    packages:
+    - python-software-properties
+    # We need gcc >= 4.8 for some c++11 features
+    - g++-4.8
```
Comment (rec) [inline .travis.yml]: The symbol `4.8` is used three times.  You might consider extracting it as another variable `GCC_VERSION` so when we upgrade to 4.9 or whatever we only have to change one place...

File: .travis.yml [hunk 3]
```diff
@@ -36,7 +47,11 @@ before_install:
   # Does it already exist? Who owns? What permissions?
   - ls -lah ~/.npm || mkdir ~/.npm
   # Make sure we own it
-  - sudo chown -R $USER ~/.npm
+  - chown -Rc $USER ~/.npm
+  # Build boost ourselves. Maintenance note: to move to a new version
+  # of boost, update BOOST_ROOT in the env: section, and the
+  # URL and paths used in the bin/sh/install-boost.sh script.
+  - bash bin/sh/install-boost.sh
```
Comment (rec) [inline .travis.yml]: Shouldn't we be trying to build static libraries by default?  that's our policy in SConstruct.

File: bin/sh/install-boost.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+set -e
+if [ ! -d "$BOOST_ROOT/lib" ]
+then
```
Comment (rec) [inline bin/sh/install-boost.sh]: Consider again extracting out the version number `1_59_0` so we can upgrade and change just in one place!  :-)

**Comment by rec** [comment]
The advantage to the changes I proposed is that you can put the things that need to be customized at the very top two lines in the file with a comment, and then "the next guy" doesn't have to read the whole file but just change one of the top two lines!

**Comment by ximinez** [comment]
@rec I agree, and was planning to make these changes when I pop enough items off of my stack.

File: .travis.yml [hunk 5]
```diff
@@ -45,7 +60,7 @@ script:
   - scons vcxproj
   - git diff --exit-code
   # $CC will be either `clang` or `gcc` (If only we could do -j12 ;)
-  - scons $CC.$TARGET
+  - scons -j2 $CC.$TARGET
```
Comment (sublimator) [inline .travis.yml]: Great!

File: .travis.yml (hunk 3 above)
Comment (ximinez) [inline .travis.yml]: This script does build static libraries, AFAIK.

**Comment by ximinez** [comment]
@rec I was able to get all the customizations into the environment except the `g++-4.8` package name. The var name didn't work there, and I don't see any incantation that would enable it.

**Comment by vinniefalco** [comment]
"Duration: 1 hr 28 min 56 sec" this is disappointing...

**Comment by ximinez** [comment]
@vinniefalco Where are you seeing that time report? The latest build, which was able to take advantage of the cache, ran for 52 minutes overall, but the longest individual build was under 20 minutes. (I assume the difference was startup time.)  https://travis-ci.org/ripple/rippled/builds/79913498

That compares quite favorably to the build for the tip of develop: https://travis-ci.org/ripple/rippled/builds/80146646. That one "ran for 1 hr 32 min 45 sec", and the longest individual build was over 41 minutes.

There is also certainly a possibility that we can optimize the caching with future changes to make the builds run even faster.

Another difference that I didn't point out is that containers allow builds to run longer than 50 minutes, where we've recently had several non-container builds have spurious failures when they're killed after 50 minutes.

**Comment by rec** [comment]
Well, I'm willing to :+1: this and then look into new ways to speed up.

Moving to the new Travis is a prerequisite for any further development so let's just do it.

**Pull request merged**